src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs

```diff
@@ -955,8 +955,8 @@ internal override void Initialize()
                 }
 
                 // create arrays for widths and alignment
-                int[] columnWidths = new int[columns];
-                int[] alignment = new int[columns];
+                Span<int> columnWidths = stackalloc int[columns];
+                Span<int> alignment = stackalloc int[columns];
```

Comment:
When it comes to using `stackalloc`, I think there are two things we need to be careful:
1. It's OK to avoid allocating heap memory when we just need small sizes, but we need to be sure `length` is small before using `stackalloc int[length]`
2. memory allocated with `stackalloc` should be used for operations that only happen in the current stack frame. The memory is associated with a given stack frame, so it can be dangerous to pass it around, because that may enable referring to memory that's no longer valid (e,g. it's passed to another method as parameter argument, and then gets returned from that method).

Here, I see `columnWidths` and `alignment` are passed to another method `Initialize`. This is not a safe usage of `stackalloc`. Plus, `Writer.Initialize` takes array arguments, so the `Span<int>` will be converted to array again.

Reply:
I agree about Length. Will review.

I don't understand why stackalloc is unsafe - we use Span<T> - it is just introduced to exclude unsafe code and unify working with heap and stack buffers. It is strong typed and has strong boundary checks. Roslyn and runtime do many checks to achieve this.

Reply:
The `Span` struct is just referencing the memory, in this case, the memory is created on the stack. See [this article](https://msdn.microsoft.com/en-us/magazine/mt814808.aspx), and I quote the words about "Span usage validation":

> Span usage validation. Because spans can refer to data that might be associated with a given stack frame, it can be dangerous to pass spans around in a way that might enable referring to memory that’s no longer valid.

Reply:
And below:
>Thankfully the C# compiler detects such invalid usage with ref structs and fails the compilation with an error:

Reply:
I see CoreFX use the protection in most cases. [Max threshold](https://github.com/dotnet/corefx/blob/e5bcd9aec536946e8053474475a0994436b4d5f4/src/Common/src/CoreLib/System/IO/Path.Windows.cs#L131) I found is 260. And sometimes they allocates 512 byte buffers.
Also I found [one sample](https://github.com/dotnet/corefx/blob/103639b6ff5aa6ab6097f70732530e411817f09b/src/Common/src/CoreLib/System/Globalization/CultureData.cs#L639) without protection. Seems they trust the fact that Culture name is short.

Can we assume that the number of columns is always small (<256, <512)? If no what is good threshold?

Reply:
The edge case is that `psobject` have up to `int.MaxValue` properties and hence columns.

Usually we accept that the width of the console is less than 120 so we can use this value as threshold.